engine/lib/phx/src/engine/event_bus.rs

```diff
@@ -0,0 +1,303 @@
+    next_tunnel_id: AtomicU32,
+    max_priority_locked: bool,
+}
+
+impl EventBus {
+    pub fn new() -> Self {
+        Self {
+            next_tunnel_id: AtomicU32::new(0),
+            update_pass_map: HashMap::new(),
+            events: HashMap::new(),
+            max_priority_locked: false,
+        }
+    }
+
+    pub fn lock_max_priority(&mut self) {
+        self.max_priority_locked = true;
+    }
+
+    pub fn dispatch(&mut self, update_pass: UpdatePass, lua: &mut Ref<Lua>) {
+        //* Can we handle the lua state differently? */
+        //* Maybe also with future potential multiple lua states in mind */
+        let globals = lua.globals();
+        let event_tunnels: Table = globals.get("EventTunnels").expect("Unknown table");
+
+        if let Some(message_heap) = self.update_pass_map.get_mut(&update_pass) {
+            let mut message_requests: Vec<_> = message_heap.drain().collect();
+            message_requests.sort_by(|a, b| b.priority.cmp(&a.priority));
+
+            for message_request in &mut message_requests {
+                if let Some(event) = self.events.get(&message_request.name) {
+                    for subscriber in &event.subscribers {
+                        let id = subscriber.tunnel_id;
+
+                        if !message_request.completed && !message_request.stay_alive {
+                            println!(
+                                "Dispatched event {} through tunnel with id {}",
+                                message_request.name, id
+                            )
+                        }
+
+                        // todo: only dispatch to subscriber with set entity_id if defined
+                        let tunnel_func: Function = event_tunnels
+                            .get(id)
+                            .expect(&format!("Unknown tunnel with id: {}", id));
+
+                        if let Err(e) = tunnel_func.call::<_, ()>(()) {
+                            trace!("{}", e);
+                        } else {
+                            message_request.completed = true; // Mark as completed
+                        }
+                    }
+                } else {
+                    panic!("Event not found: {}", message_request.name);
+                }
+            }
+
+            for message_request in message_requests {
+                if message_request.stay_alive || !message_request.completed {
+                    message_heap.push(message_request);
+                }
+            }
+        }
+    }
+
+    pub fn dispatch_all(&mut self, mut lua: Ref<Lua>) {
+        for update_pass in UpdatePass::iter() {
+            self.dispatch(update_pass, lua.borrow_mut());
+        }
+    }
+}
+
+#[luajit_ffi_gen::luajit_ffi]
+impl EventBus {
+    pub fn register(
+        &mut self,
+        event_name: String,
+        priority: Option<u16>, //* how do i handle Options via ffi? It requires a uint16 const pointer */
```

Comment:
Yeah, a pointer is really the only way to express an optional type across the FFI boundary, because it allows both any valid u16 value _or_ the absence of a value.

Underneath on line 156, you literally convert it from Option<u16> to u16 with `unwrap_or(0)`, so effectively `None` and `Some(0)` are exactly the same. You might as well change this to a u16 and document it that 0 is the default.